File: ui/src/shared/utils/displayNameHelpers.ts
import { UserProfile } from "@/shared/models/userProfile";

interface EmployeeDisplayDetails {
  email?: string;
  firstName?: string;
  lastName?: string;
  position?: string;
}

export const unknownEmployeeDisplayName = "Unknown employee";

export const normalizeEmailForDisplay = (email?: string): string => email?.trim().toLowerCase() ?? "";

export const stripInlineEmail = (displayName?: string | null): string =>
  displayName?.replace(/\s+\([^()\s]+@[^()\s]+\)\s*$/, "").trim() ?? "";

export const getUserProfileName = (profile: UserProfile | null, fallback = "User"): string =>
  [profile?.firstName, profile?.lastName].filter(Boolean).join(" ").trim() || fallback;

export const getUserDisplayName = (profile: UserProfile | null, email?: string): string => {
  const normalizedEmail = normalizeEmailForDisplay(email);
  const name = getUserProfileName(profile, "");
  return name ? `${name} (${normalizedEmail})` : normalizedEmail;
};

export const getEmployeeDisplayName = (employee?: EmployeeDisplayDetails | null): string => {
  const { firstName, lastName, position, email } = employee ?? {};
  const name = [firstName, lastName]
    .map((value) => value?.trim())
    .filter((value): value is string => Boolean(value))
    .join(" ");
  const title = position?.trim();
  const identity = title ? (name ? `${name}, ${title}` : title) : name;
  return identity || normalizeEmailForDisplay(email) || unknownEmployeeDisplayName;
};

File: api/src/Click2Approve.Application/Helpers/DisplayNameHelpers.cs
namespace Click2Approve.Application.Helpers;

/// <summary>
/// Provides shared user display-name formatting.
/// </summary>
public static class DisplayNameHelpers
{
    /// <summary>
    /// Formats a participant display name from optional employee details.
    /// </summary>
    public static string FormatParticipantDisplayName(
        string? firstName,
        string? lastName,
        string? position,
        string? email)
    {
        var name = JoinNonEmpty(firstName, lastName);
        var title = position?.Trim();
        var identity = string.IsNullOrWhiteSpace(title)
            ? name
            : string.IsNullOrWhiteSpace(name)
                ? title
                : $"{name}, {title}";

        if (!string.IsNullOrWhiteSpace(identity)) return identity;

        var normalizedEmail = NormalizeEmailForDisplay(email);
        return normalizedEmail ?? throw new ArgumentException("Email must be provided if no name or title is available.", nameof(email));
    }

    /// <summary>
    /// Normalizes an email address for display labels.
    /// </summary>
    public static string NormalizeEmailForDisplay(string? email)
    {
        return email?.Trim().ToLowerInvariant() ?? string.Empty;
    }

    /// <summary>
    /// Joins non-empty text parts with a single space.
    /// </summary>
    public static string JoinNonEmpty(params string?[] values)
    {
        return string.Join(" ", values
            .Select(value => value?.Trim())
            .Where(value => !string.IsNullOrWhiteSpace(value)));
    }
}
